amr-wind/physics/ForestDrag.cpp

```diff
@@ -0,0 +1,169 @@
+            m_y_forest.size());
+        amrex::Gpu::DeviceVector<amrex::Real> device_ht_forest(
+            m_ht_forest.size());
+        amrex::Gpu::DeviceVector<amrex::Real> device_dia_forest(
+            m_dia_forest.size());
+        amrex::Gpu::DeviceVector<amrex::Real> device_cd_forest(
+            m_cd_forest.size());
+        amrex::Gpu::DeviceVector<amrex::Real> device_lai_forest(
+            m_lai_forest.size());
+        amrex::Gpu::DeviceVector<amrex::Real> device_lm_forest(
+            m_lm_forest.size());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_type_forest.begin(),
+            m_type_forest.end(), device_type_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_x_forest.begin(), m_x_forest.end(),
+            device_x_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_y_forest.begin(), m_y_forest.end(),
+            device_y_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_ht_forest.begin(), m_ht_forest.end(),
+            device_ht_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_dia_forest.begin(), m_dia_forest.end(),
+            device_dia_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_cd_forest.begin(), m_cd_forest.end(),
+            device_cd_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_lai_forest.begin(), m_lai_forest.end(),
+            device_lai_forest.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_lm_forest.begin(), m_lm_forest.end(),
+            device_lm_forest.begin());
+        const auto* type_forest_ptr = device_type_forest.data();
+        const auto* x_forest_ptr = device_x_forest.data();
+        const auto* y_forest_ptr = device_y_forest.data();
+        const auto* ht_forest_ptr = device_ht_forest.data();
+        const auto* dia_forest_ptr = device_dia_forest.data();
+        const auto* cd_forest_ptr = device_cd_forest.data();
+        const auto* lai_forest_ptr = device_lai_forest.data();
+        const auto* lm_forest_ptr = device_lm_forest.data();
+        const auto& phy_mgr = m_sim.physics_manager();
+        for (amrex::MFIter mfi(velocity); mfi.isValid(); ++mfi) {
+            const auto& vbx = mfi.validbox();
+            auto levelDrag = drag.array(mfi);
+            const unsigned forestSize = m_x_forest.size();
+            amrex::ParallelFor(
+                vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    // compute the source term
+                    const amrex::Real x = prob_lo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real y = prob_lo[1] + (j + 0.5) * dx[1];
+                    const amrex::Real z = prob_lo[2] + (k + 0.5) * dx[2];
+                    levelDrag(i, j, k) = 0.0;
+                    for (unsigned ii = 0; ii < forestSize; ++ii) {
```

Comment:
I don't like this for loop combined with the while loop below. Let's chat and see if we can come up with something better.